### ISSUE TITLE
Masks

### ISSUE THREAD

**Issue by lemuriandezapada** [fchollet/keras#370]
From what I understand so far, masks "hide" a particular input or output data item so they don't get updated or contribute to learning. Which is obviously awesome. 
However it seems to only work on -items-. Any chance to get it to work on points (as in if you have multidimensional time input for example but missing particular items from that dimension)? Would be a good way to deal with missing features and/or outputs.

**Comment by beka-evature** [comment]
I'm not sure what you've meant with items vs. points, but I think some of the answers lie in #355

**Comment by fchollet** [comment]
> masks "hide" a particular input or output data item

Specific typesteps in a temporal input, precisely.

>  Any chance to get it to work on points (as in if you have multidimensional time input for example but missing particular items from that dimension)? Would be a good way to deal with missing features and/or outputs.

There are no plans to implement this. But if you need it for a specific type of layer (like Dense) it would be easy for you to re-implement a version of that layer that supports feature-wise masking.

**Comment by wxs** [comment]
I might be missing something, but for non-temporal data can't you just set your masked values to 0?

The reason it's more complex in the temporal case is because of the recurrent connections "introducing" data (via biases) for unseen timesteps.

**Comment by lemuriandezapada** [comment]
Not sure about that. Setting values to 0 actually gives them 0 value and.
For example in a feedforward case, you want to learn a mapping from an input data to a target multi-d output, say: 
1 2 3 4 2 u 1 4 5 
2 u 1 2 4 5 u 2 1
2 3 4 u 5 2 4 5 1 
Where u stands for datapoints you don't know. You'll compute gradients for these. If you set your unknown value to 0 then it'll compute the gradient between your target value (0) and the resulting one from the network (whatever) and backprop that. Which would have you learn crap. What you'd want would be that for the values you're missing to just ignore the gradient and just compute from it from the rest of the data.

For the input it's a similar manner. Its not just that you're making them go away if you're setting them to 0, you're actually telling your network to map FROM input unit = 0. This might work-ish if your data is normally distributed, but otherwise you're still introducing wrong data, for example, also via biases. It's still fine-ish when you compute them, but when you propagate the errors you will still update the bias to learn to map from the 0 input value. Which is again wrong. But you're right, this would involve a bit of weight scaling to patch up.

**Comment by wxs** [comment]
I think you're talking about masking the *input* in which case the following argument applies. Unless you're talking about the output? That's a completely different discussion.

Here is my argument for why you don't "learn crap": OK so one of my input data points is 0 (masked), going into a dense layer. There are a number of weights coming out of that data point. The first hidden layer is given by

    h0_i = activation(b_i + sum_j x_j W_ij)

Say x_j' is 0 (masked), then

    dErr/dW_ij' = dErr/dh0_i * dh0_i / dW_ij'

The first term in that product `dErr/dh0_i` is a function of the higher levels of the model. The second term

    dh0_i / dW_ij' = d activation/dt * x_j'

but remember that x_j' is 0, so the derivative of the error with respect to any of the weights coming out of x_j' (W_ij') is 0, so the gradient step will *not* update those weights!

I'm pretty sure this argument is sound.

**Comment by wxs** [comment]
Now *output* masking on the other hand, with "don't care" values, that gets interesting. If I'm not mistaken, every objective function would need to support the masking concept, and not include masked values in its calculation. And you would need to pass a mask in with your `y`, or else define special masked_value on `y`.

**Comment by lemuriandezapada** [comment]
The gradient step will not update to those weights, but they will update to every other weight for every other datapoint in the context in which that input was 0. Your lack of input will give you the wrong error at the end of the network to propagate back. This will affect all the other weights in a way inconsistent with the real data. 

Let's assume you're missing (and thus masking to 0) all datapoints. None of the data weights will change but your bias will change to compensate for the fact that all your inputs are 0. Which they're not actually 0, they're just unknown and your bias shouldn't move as much. :P 

Another way you can look at this is that by setting the inputs to 0, you're effectively doing dropout but without scaling up the rest of the weights to compensate (which I actually haven't checked to see if keras is doing :)) ). I've done it before and it worked wonderfully, though I'm not sure how it would be implemented in Keras.

Output masking on the other hand would be much easier to do and much more useful though, indeed.

**Comment by wxs** [comment]
OK that argument kind of makes sense to me, though I don't know what the correct way to alter the update equations is, then.

It seems like in that case the mathematically correct solution to, given some prior over the masked input dimension, integrate your update across all values of that input? If so, what is the correct way to determine that prior?

**Comment by lemuriandezapada** [comment]
I often just scale the remaining weights (or input values) including the bias, by the sum of all the weights/the sum of the remaining weights, then backprop as usual. You effectively renormalize the inputs to compensate for the missing value. This worked well for me and was able to reconstruct missing values which would not usually be possible with just setting inputs to 0 because you effectively take out input values. Though I'm sure there's a smarter way to do this. 

Then again I could just do a hacky version of this on the inputs myself before feeding them to the network at each run and should work ok. 

But the output masking would be a nice feature, not sure how I would implement it, though. Need a change in the loss function at each iteration depending on a mask.

**Comment by wxs** [comment]
> But the output masking would be a nice feature, not sure how I would implement it, though. Need a change in the loss function at each iteration depending on a mask.

You can do this now with `sample_weight` I think. Just pass in your mask as the `sample_weight` parameter to `fit()`

**Comment by lemuriandezapada** [comment]
beautiful, thank you :)

**Comment by oakkas** [comment]
In the case of sample_weight how can we give a different weight for each example. As in the example above:
1 2 3 4 2 u 1 4 5
2 u 1 2 4 5 u 2 1
2 3 4 u 5 2 4 5 1

the missing value would be in a different index and how to provide this as sample_weight?
